GridKit/Model/PowerElectronics/SystemModelPowerElectronics.hpp

```diff
@@ -123,18 +127,140 @@ namespace GridKit
     {
       for (auto comp : this->components_)
         delete comp;
+
+      if (jac_sparsity_row_indices != nullptr)
+      {
+        delete[] jac_sparsity_row_indices;
+      }
+      if (jac_sparsity_col_indices != nullptr)
+      {
+        delete[] jac_sparsity_col_indices;
+      }
     }
 
     /**
-     * @brief allocator default
-     *
-     * @todo this should throw an exception as no allocation without a graph is allowed.
-     * Or needs to be removed from the base class
-     *
-     * @return int
+     * @brief Allocates and constructs sparsity patterns for system jacobian.
+     *        To do this, sparsity patterns of components are needed, so
+     *        each component's jacobian is evaluated, and the sparsity pattern
+     *        of that component's jacobian is expected to match the sparsity pattern
+     *        for the rest of the simulation.
+     * @pre   `size_` must be correctly initialized
      */
     int allocate()
     {
+      // Helper struct for identifying a particular component's local system variable
+      struct ComponentContribution
+      {
+        // The index of the component in `components_`.
+        size_t comp_idx_;
+        // The local system variable index
+        size_t local_row_idx_;
+      };
+
+      // Helper struct for saving CSR sparsities of a particular component, since
+      // right now they only report COO. Can be removed once components have CSR matrices.
+      struct CsrSparsity
+      {
+        std::vector<IdxT> row_indices_;
+        std::vector<IdxT> col_indices_;
```

Comment:
I would avoid using `std::vector` here. We created CSR matrix class to eliminate need for STL containers. We can use the CSR matrix here instead.